src/cockpit/samples.py

```diff
@@ -192,23 +202,25 @@ class CGroupSampler(Sampler):
     cgroups_v2: Optional[bool] = None
     cgroups_v2_path = '/sys/fs/cgroup/'
 
-    def read_cgroup_keyed_stat(self, samples, path, cgroup, name, key):
-        with open(path) as stat:
-            for line in stat:
+    def read_cgroup_keyed_stat(self, samples, rootfd, statfile, cgroup, name, key):
+        with os_open(statfile, os.O_RDONLY, dir_fd=rootfd) as fd:
+            stat = os.read(fd, 1024)
+            for line in stat.decode().splitlines():
```

Comment:
Why decode?  I think this is actually kinda bogus.  The kernel interface is definitely expressed in locale-independent bytes.  You should rather change all of the key values to bytestrings, and then just work in bytes.

Reply:
Also: I wonder if we could do better than the splitlines approach.  Maybe directly scan for the key, using `.index()`?

Reply:
Try to think in terms of memory allocations to get a rough idea of what takes time and not — converting bytes to a unicode object is a memory allocation.  Splitting a file into multiple lines is a memory allocation for each line, plus at least one for the array.  Also: working on unicode objects is potentially slower than working on raw bytes.

Reply:
Yeah, the decode() was a quick hack from late Friday afternoon. I'll keep it at bytes. Seems to be slightly faster:

Old
```
In [41]: !hyperfine --warmup 3 'python3 bench_sampler.py'
Benchmark 1: python3 bench_sampler.py
  Time (mean ± σ):      2.407 s ±  0.440 s    [User: 1.470 s, System: 0.906 s]
  Range (min … max):    1.973 s …  3.257 s    10 runs
```

New
```
In [44]: !hyperfine --warmup 3 'python3 bench_sampler.py'
Benchmark 1: python3 bench_sampler.py
  Time (mean ± σ):      2.095 s ±  0.128 s    [User: 1.258 s, System: 0.809 s]
  Range (min … max):    1.964 s …  2.331 s    10 runs
```